### PR TITLE
LibCpp: Bunch of parser improvements

### PR DESCRIPTION
**Pull request by alimpfard**
Most of these basically ignore things, as we don't really have the means to represent them (e.g. `virtual`, or `const`-ness of a member function, or inherited classes, etc), so I didn't include any tests.
cc @itamar8910, if you want, I could add tests for the things that actually have an AST representation (e.g. reference types, east const, function types)

### CODE REVIEW

File: Userland/Libraries/LibCpp/AST.h [hunk 1]
```diff
@@ -281,6 +281,33 @@ class Pointer : public Type {
...
+    {
+    }
+
+    const Type* pointee() const { return m_pointee.ptr(); }
+    void set_pointee(RefPtr<Type>&& pointee) { m_pointee = move(pointee); }
+    Kind kind() const { return m_kind; }
+
+private:
+    RefPtr<Type> m_pointee;
+    Kind m_kind;
```
Comment (itamar8910) [inline Userland/Libraries/LibCpp/AST.h:308]: Since it's only set in the c'tor, can be made const.

File: Userland/Libraries/LibCpp/Parser.cpp [hunk 2]
```diff
@@ -1237,6 +1237,16 @@ NonnullRefPtr<Type> Parser::parse_type(ASTNode& parent)
         type = ptr;
     }
 
+    if (!eof() && (peek().type() == Token::Type::And || peek().type() == Token::Type::AndAnd)) {
+        type->set_end(position());
+        auto ref = consume();
+        auto ptr = create_ast_node<Reference>(parent, type->start(), ref.end(), ref.type() == Token::Type::And ? Reference::Kind::Lvalue : Reference::Kind::Rvalue);
```
Comment (itamar8910) [inline Userland/Libraries/LibCpp/Parser.cpp:1269]: nitpick: `ptr`=>`ref` (and `ref`=>`ref_token` or something :sweat_smile: : )

**Review by itamar8910 — APPROVED** [review]
Awesome improvements! It's so nice seeing other people taking interest in working on LibCpp :)

I like the approach you took here for dealing with escaped newlines. I had a local branch that did string concatenation to work around it, but tweaking the lexer as you did here is defn better :+1:

File: Userland/Libraries/LibCpp/AST.h (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibCpp/AST.h:308]: I assume you mean `RefPtr<Type const>`?

File: Userland/Libraries/LibCpp/Parser.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibCpp/Parser.cpp:1269]: Sure!

**Comment by alimpfard** [comment]
> It's so nice seeing other people taking interest in working on LibCpp

^^' I'm doing this out of pure self-interest, the more stuff LibCpp can parse correctly, the more stuff I can do with my (in progress, not yet even close to done :P) automatic JS binding generator.

File: Userland/Libraries/LibCpp/AST.h (hunk 1 above)
Comment (itamar8910) [reply Userland/Libraries/LibCpp/AST.h:308]: I was actually referring to `m_kind`.

Comment (alimpfard) [reply Userland/Libraries/LibCpp/AST.h:308]: oh
eh, I don't think making that `const` makes any difference, it's just a private enum member

**Pull request merged**